src/main.rs

```diff
@@ -0,0 +1,160 @@
+use std::{
+    collections::HashMap,
+    env, fs,
+    time::{Duration, Instant},
+};
+
+use serde::{Deserialize, Serialize};
+use spectre::{graph::AGraph, graph::Graph};
+
+#[derive(Default, Clone, Serialize, Deserialize)]
+pub struct Node {
+    ip: String,
+    betweenness: f64,
+    closeness: f64,
+    num_connections: usize,
+}
+
+#[derive(Default, Clone, Serialize, Deserialize)]
+pub struct CrunchyState {
+    agraph_length: u32,
+    elapsed: f64,
+    nodes: Vec<Node>,
+    min_betweenness: f64,
+    max_betweenness: f64,
+    min_closeness: f64,
+    max_closeness: f64,
+}
+
+#[allow(dead_code)]
+#[derive(Default, Clone, Deserialize)]
+pub struct NetworkSummary {
+    num_known_nodes: usize,
+    num_good_nodes: usize,
+    num_known_connections: usize,
+    num_versions: usize,
+    protocol_versions: HashMap<u32, usize>,
+    user_agents: HashMap<String, usize>,
+    crawler_runtime: Duration,
+    node_ips: Vec<String>,
+    agraph: AGraph,
+}
+
+#[allow(dead_code)]
+#[derive(Default, Clone, Deserialize)]
+pub struct JsonRpcResponse {
+    jsonrpc: String,
+    result: NetworkSummary,
+    id: usize,
+}
+
+#[derive(Default, Clone, Deserialize)]
+pub struct ResponseSample {
+    pub response: JsonRpcResponse,
+}
+
+pub fn load_response(filepath: &str) -> JsonRpcResponse {
+    let jstring = fs::read_to_string(filepath).unwrap();
+    let response: JsonRpcResponse = serde_json::from_str(&jstring).unwrap();
+    response
+}
+
+pub fn load_state(filepath: &str) -> CrunchyState {
+    let jstring = fs::read_to_string(filepath).unwrap();
+    let response: CrunchyState = serde_json::from_str(&jstring).unwrap();
+    response
+}
+
+fn compute_betweenness_and_closeness(infile: &str, outfile: &str) {
+    let response = load_response(infile);
+    let agraph = response.result.agraph;
+    let graph: Graph<usize> = Graph::new();
+    let start = Instant::now();
+    let (betweenness, closeness) = graph.compute_betweenness_and_closeness(&agraph);
+    let elapsed = start.elapsed();
+    let mut min_betweenness: f64 = 10000.0;
+    let mut max_betweenness: f64 = 0.0;
+    let mut min_closeness: f64 = 10000.0;
+    let mut max_closeness: f64 = 0.0;
+    let mut num_connections = vec![0; agraph.len()];
+    for (n, connections) in agraph.iter().enumerate() {
+        num_connections[n] = connections.len();
+    }
```

Comment:
```suggestion
    let num_connections = agraph.iter().map(|connections| connections.len()).collect::<Vec<_>>()`
```
(haven't tested, but it should be more or less it)

Reply:
done.  That code is now gone -- array was no longer used.